File: mobile/lib/models/schedule.dart
class Schedule {
  final String id;
  final String title;
  final String? description;
  final DateTime startTime;
  final DateTime? endTime;
  final String? location;
  final double? latitude;
  final double? longitude;
  final String status;
  final String transportMode; // car, motorcycle, transit, bike, walk
  final List<String>? attendIds;
  final List<Map<String, dynamic>>? attends; // Added attends list
  final String? contactName;
  final String? contactEmail;
  final String? contactPhone;
  final String? contactLineId;

  final String? cancelReason;
  final bool? isOnline;
  final bool? isOwner;
  final String? creatorName;

  Schedule({
    required this.id,
    required this.title,
    this.description,
    required this.startTime,
    this.endTime,
    this.location,
    this.latitude,
    this.longitude,
    required this.status,
    required this.transportMode,
    this.attendIds,
    this.attends,
    this.cancelReason,
    this.contactName,
    this.contactEmail,
    this.contactPhone,
    this.contactLineId,
    this.isOnline,
    this.isOwner,
    this.creatorName,
  });

  factory Schedule.fromJson(Map<String, dynamic> json) {
    return Schedule(
      id: json['id'] as String? ?? '', // Handle possible mismatch
      title: json['title'] as String? ?? 'No Title', // Handle null title
      description: json['description'] as String?,
      startTime: DateTime.parse(json['start_time'] ?? json['startTime'] ?? DateTime.now().toIso8601String()),
      endTime: json['end_time'] != null
          ? DateTime.parse(json['end_time'])
          : (json['endTime'] != null ? DateTime.parse(json['endTime']) : null),
      location: json['location'] as String?,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      status: _normalizeStatus(json['status'] as String?),
      transportMode: json['transport_mode'] as String? ?? json['transportMode'] as String? ?? 'car',
      cancelReason: json['cancel_reason'] as String?,
      contactName: json['contact_name'] as String?,
      contactEmail: json['contact_email'] as String?,
      contactPhone: json['contact_phone'] as String?,
      contactLineId: json['contact_line_id'] as String?,
      attendIds: (json['attend_ids'] as List<dynamic>?)
          ?.map((e) => e.toString())
          .toList(),
      attends: (json['attends'] as List<dynamic>?)
          ?.map((e) => e as Map<String, dynamic>)
          .toList(),
      isOnline: json['is_online'] == true,
      isOwner: json['is_owner'] != false,
      creatorName: json['creator_name'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'start_time': startTime.toIso8601String(),
      'end_time': endTime?.toIso8601String(),
      'location': location,
      'latitude': latitude,
      'longitude': longitude,
      'transport_mode': transportMode,
      'attend_ids': attendIds,
      'attends': attends,
      'cancel_reason': cancelReason,
      'is_online': isOnline,
      'contact_name': contactName,
      'contact_email': contactEmail,
      'contact_phone': contactPhone,
      'contact_line_id': contactLineId,
      'is_owner': isOwner,
      'creator_name': creatorName,
    };
  }

  static String _normalizeStatus(String? status) {
    if (status == null || status == 'P') return 'PD';
    return status;
  }
}

File: server/app/models/schedule.py
from sqlmodel import SQLModel, Field
from sqlalchemy import Column, Integer, Identity, String, DateTime
from typing import Optional, List
from datetime import datetime
from ..utils.id_generator import generate_schedule_id

class Schedule(SQLModel, table=True):
    __tablename__ = "schedule"

    # id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY
    id: Optional[int] = Field(
        default=None, 
        sa_column=Column(Integer, Identity(always=True), primary_key=True)
    )
    
    # user_id VARCHAR(255) NOT NULL
    # This refers to users.user_id, NOT users.id
    user_id: str = Field(sa_column=Column(String(255), nullable=False))
    
    # schedule_id VARCHAR(255) NOT NULL
    schedule_id: str = Field(
        default_factory=generate_schedule_id,
        sa_column=Column(String(255), unique=True, nullable=False)
    )
    
    # meeting_time merged into start_time
    # meeting_end_time merged into end_time
    
    # meeting_start_time TIMESTAMPTZ
    meeting_start_time: datetime = Field(
        sa_column=Column(DateTime(timezone=True), nullable=True)
    )
    
    # meeting_end_time TIMESTAMPTZ - Mandatory
    meeting_end_time: datetime = Field(
        sa_column=Column(DateTime(timezone=True), nullable=False)
    )
    
    # meeting_location VARCHAR(255) NULL
    # Keeping meeting_location column name for compatibility if needed, or alias?
    # User only asked for start/end time. I will keep meeting_location but alias it to location property?
    # Actually, let's keep meeting_location column but expose 'location' as alias or just use meeting_location
    meeting_location: Optional[str] = Field(default=None, sa_column=Column(String(255), nullable=True))
    
    # status VARCHAR(2) DEFAULT 'PD' (= Status.PENDING; mobile only recognizes 'PD')
    status: str = Field(default="PD", sa_column=Column(String(2), nullable=True, server_default="PD"))
    
    # cancel_reason VARCHAR(255) NULL
    cancel_reason: Optional[str] = Field(default=None, sa_column=Column(String(255), nullable=True))
    
    # Contact Details
    contact_id: Optional[int] = Field(default=None, sa_column=Column(Integer, nullable=True))
    # Legacy fields removed as per refactor plan
    # contact_name, contact_email, contact_phone, contact_line_id are removed.
    
    # created_at TIMESTAMPTZ
    created_at: datetime = Field(
        default_factory=datetime.now,
        sa_column=Column(DateTime(timezone=True), nullable=True)
    )
    
    # updated_at TIMESTAMPTZ
    updated_at: datetime = Field(
        default_factory=datetime.now,
        sa_column=Column(DateTime(timezone=True), nullable=True)
    )

    # --- Extra fields for App Functionality (not in requested schema but needed) ---
    
    title: str = Field(default="No Title")
    description: Optional[str] = None
    
    transport_mode: Optional[str] = None
    type: Optional[str] = None # meeting/personal
    attends_display: Optional[str] = Field(default=None, alias="attends") # For text display
    is_reminder: bool = Field(default=False)

    # Departure-reminder support (see ReminderService / background_reminder_scheduler)
    # Computed "time to leave" based on travel duration to meeting_location.
    reminder_leave_by_time: Optional[datetime] = Field(
        default=None, sa_column=Column(DateTime(timezone=True), nullable=True)
    )
    # User-adjustable offsets (minutes). Two independent reminders:
    #  - reminder_before_start_minutes: fires N minutes before meeting_start_time
    #  - reminder_before_leave_minutes: fires N minutes before reminder_leave_by_time
    reminder_before_start_minutes: int = Field(
        default=60, sa_column=Column(Integer, nullable=True, server_default="60")
    )
    reminder_before_leave_minutes: int = Field(
        default=60, sa_column=Column(Integer, nullable=True, server_default="60")
    )

    # Coordinates for Map
    latitude: Optional[float] = Field(default=None)
    longitude: Optional[float] = Field(default=None)

    # Online event flag
    is_online: bool = Field(default=False)

    # Relationship
    from sqlmodel import Relationship
    # Use string forward reference and explicit join condition since we use custom string IDs
    attend_records: List["attend"] = Relationship(
        sa_relationship_kwargs={
            "primaryjoin": "Schedule.schedule_id==foreign(attend.schedule_id)",
            "cascade": "all, delete-orphan",
            "uselist": True
        }
    )

    contact: Optional["Contact"] = Relationship(
        sa_relationship_kwargs={
            "primaryjoin": "Schedule.contact_id==foreign(Contact.id)",
            "uselist": False
        }
    )
    
    # Mappings
    @property
    def location(self) -> Optional[str]:
        """Alias for meeting_location for app compatibility"""
        return self.meeting_location
    
    @location.setter
    def location(self, value: Optional[str]):
        self.meeting_location = value
    
    # Override dict() to include computed properties
    def dict(self, *args, **kwargs):
        """Override dict to include @property fields for API serialization"""
        data = super().dict(*args, **kwargs)
        # Add computed properties that the frontend expects
        data['start_time'] = self.meeting_start_time.isoformat() if isinstance(self.meeting_start_time, datetime) else self.meeting_start_time
        data['end_time'] = self.meeting_end_time.isoformat() if isinstance(self.meeting_end_time, datetime) else self.meeting_end_time
        data['location'] = self.meeting_location
        data['latitude'] = self.latitude
        data['longitude'] = self.longitude
        data['id'] = self.schedule_id  # Frontend expects 'id' to be the schedule_id
        data['is_online'] = self.is_online
        
        # Serialize attends
        try:
            data['attends'] = [a.dict() for a in self.attend_records]
        except Exception as e:
            data['attends'] = []
            
        # Serialize contact info from relationship if available
        if self.contact_id and self.contact:
            data['contact_name'] = self.contact.nick_name  # Contact has nick_name
            data['contact_email'] = self.contact.email
            data['contact_phone'] = self.contact.phone
            data['contact_line_id'] = self.contact.line_id
            
        return data
